dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+
+    if (IsEqualIID(riid, IID_IUnknown) || IsEqualIID(riid, IID_IRecycleBinEnumList))
+        *ppvObject = static_cast<IRecycleBinEnumList *>(this);
+    else
+    {
+        *ppvObject = NULL;
+        return E_NOINTERFACE;
+    }
+
+    AddRef();
+    return S_OK;
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::AddRef()
+{
+    TRACE("(%p)\n", this);
+    return InterlockedIncrement(&m_ref);
+}
+
+RecycleBin5Enum::~RecycleBin5Enum()
+{
+    TRACE("(%p)\n", this);
+
+    m_recycleBin->OnClosing(this);
+    UnmapViewOfFile(m_pInfo);
+    m_recycleBin->Release();
+    SHFree(m_pszPrefix);
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::Release()
+{
+    TRACE("(%p)\n", this);
+
+    ULONG refCount = InterlockedDecrement(&m_ref);
+    if (refCount == 0)
+        delete this;
+    return refCount;
+}
+
+STDMETHODIMP RecycleBin5Enum::Next(DWORD celt, IRecycleBinFile **rgelt, DWORD *pceltFetched)
+{
+    HRESULT hr;
+
+    TRACE("(%p, %u, %p, %p)\n", this, celt, rgelt, pceltFetched);
+
+    if (!rgelt)
+        return E_POINTER;
+    if (!pceltFetched && celt > 1)
+        return E_INVALIDARG;
+
+    ULARGE_INTEGER FileSize;
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
+    if (FileSize.u.LowPart == 0)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    DWORD dwEntries =
+        (DWORD)((FileSize.QuadPart - sizeof(INFO2_HEADER)) / sizeof(DELETED_FILE_RECORD));
+
+    DWORD iEntry = m_dwCurrent, fetched = 0;
+    DELETED_FILE_RECORD *pDeletedFile = (DELETED_FILE_RECORD *)(m_pInfo + 1) + iEntry;
+    for (; iEntry < dwEntries && fetched < celt; ++iEntry)
+    {
+        hr = RecycleBin5File_Constructor(m_recycleBin, m_pszPrefix, pDeletedFile, &rgelt[fetched]);
+        if (SUCCEEDED(hr))
+            fetched++;
+        pDeletedFile++;
+    }
+
+    m_dwCurrent = iEntry;
+    if (pceltFetched)
+        *pceltFetched = fetched;
+    if (fetched == celt)
+        return S_OK;
+    else
+        return S_FALSE;
+}
+
+STDMETHODIMP RecycleBin5Enum::Skip(DWORD celt)
+{
+    TRACE("(%p, %u)\n", this, celt);
+    m_dwCurrent += celt;
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5Enum::Reset()
+{
+    TRACE("(%p)\n", this);
+    m_dwCurrent = 0;
+    return S_OK;
+}
+
+RecycleBin5Enum::RecycleBin5Enum(
+    IN IRecycleBin5 *prb,
+    IN HANDLE hInfo,
+    IN OUT INFO2_HEADER *pInfo,
+    IN OUT LPWSTR pszPrefix)
+    : m_ref(1)
+    , m_recycleBin(prb)
+    , m_hInfo(hInfo)
+    , m_pInfo(pInfo)
+    , m_dwCurrent(0)
+    , m_pszPrefix(pszPrefix)
+{
+    m_recycleBin->AddRef();
+}
+
+EXTERN_C
+HRESULT
+RecycleBin5Enum_Constructor(
+    IN IRecycleBin5 *prb,
+    IN HANDLE hInfo,
+    IN HANDLE hInfoMapped,
+    IN LPCWSTR szPrefix,
+    OUT IUnknown **ppUnknown)
+{
+    if (!ppUnknown)
+        return E_POINTER;
+
+    LPWSTR pszPrefix;
+    HRESULT hr = SHStrDup(szPrefix, &pszPrefix);
+    if (FAILED(hr))
+        return hr;
+
+    INFO2_HEADER *pInfo = (INFO2_HEADER *)MapViewOfFile(hInfoMapped, FILE_MAP_READ, 0, 0, 0);
```

Comment:
I think this whole pre-init code could be put in the RecycleBin5Enum C++ constructor.

Because as it currently is, it's RecycleBin5Enum_Constructor(), not belonging to the class, that calls the MapView..., but it's the `~RecycleBin5Enum()` class destructor that calls the UnMap... routine (+ the other CloseHandles calls).

Reply:
Used `Init()` helper method.